search/search-client-query-transformer/src/main/java/io/camunda/search/clients/DocumentBasedSearchClients.java

```diff
@@ -182,9 +190,77 @@ public List<ProcessDefinitionFlowNodeStatisticsEntity> processDefinitionFlowNode
   @Override
   public SearchQueryResult<ProcessInstanceEntity> searchProcessInstances(
       final ProcessInstanceQuery filter) {
+    if (!filter.filter().incidentErrorHashCodes().isEmpty()) {
+      return searchProcessInstancesByIncidentErrorHash(filter);
+    }
     return getSearchExecutor().search(filter, ProcessInstanceForListViewEntity.class);
   }
 
+  private SearchQueryResult<ProcessInstanceEntity> searchProcessInstancesByIncidentErrorHash(
+      final ProcessInstanceQuery filter) {
+
+    final var originalFilter = filter.filter();
+
+    final var incidentFilter =
+        FilterBuilders.incident(
+            f ->
+                f.errorMessageHashes(originalFilter.incidentErrorHashCodes())
+                    .states(IncidentState.ACTIVE));
+
+    final var incidentResult = searchIncidents(IncidentQuery.of(f -> f.filter(incidentFilter)));
+
+    if (incidentResult.items().isEmpty()) {
+      return new SearchQueryResult.Builder<ProcessInstanceEntity>().build();
+    }
+
+    final Set<Long> processInstanceKeys = new HashSet<>();
+    incidentResult.items().forEach(i -> processInstanceKeys.add(i.processInstanceKey()));
+
+    for (final var op : originalFilter.processInstanceKeyOperations()) {
+      if (op.operator().equals(Operator.EQUALS)) {
+        processInstanceKeys.add(op.value());
+      } else if (op.operator().equals(Operator.IN)) {
+        processInstanceKeys.addAll(op.values());
+      }
+    }
+
+    final var updatedFilter =
+        builderFrom(originalFilter)
+            .processInstanceKeyOperations(Operation.in(List.copyOf(processInstanceKeys)))
+            .hasIncident(true)
+            .build();
+
+    final var updatedQuery =
+        ProcessInstanceQuery.of(
+            q ->
+                q.filter(updatedFilter)
+                    .sort(filter.sort())
+                    .page(filter.page())
+                    .resultConfig(filter.resultConfig()));
+
+    return getSearchExecutor().search(updatedQuery, ProcessInstanceForListViewEntity.class);
+  }
+
+  private static Builder builderFrom(final ProcessInstanceFilter original) {
```

Comment:
🔧 Let's keep this in the `ProcessInstanceFilter` class, I think it fits better there. 

🔧 And I think usually we call the method `toBuilder`.
Update: There is already one in the `ProcessInstanceFilter`!


```suggestion
  private static Builder toBuilder(final ProcessInstanceFilter original) {
```

Reply:
@danielkelemen The only reason I moved it away from the `ProcessInstanceFilter` class was because it was not copying all the fields, it excludes the `processInstanceKey' because that will be updated by the calling method. I thought this will make it too specific to be reused.

Reply:
I noticed after my review (the 'update' part) that there is already a `toBuilder` method in the `ProcessInstanceFilter`. Let's use that instead of this new method.
That should work the same way if I'm not mistaken.

Reply:
I have reused the `toBuilder` method as you suggested and introduced the `replaceProcessInstanceKeyOperations` method to handle the replacement of the field.